### DISCUSSION TITLE
Efficiently read the dictionary every time the IME is activated

### DISCUSSION THREAD

**Discussion by Rohupt** [saschanaz/ime-rs#11] [General]
The dictionary is quite large now, and I had to transform it into a HashMap upon reading from file for faster wordfinding. However, it's quite troublesome that every time I switch out of and into the IME again, the DLL reread it (it takes quite a few seconds for the IME to start working, which is not ideal). So I'm thinking of somehow store the HashMap dictionary to a separate file, and have the DLL read from it rather than from a human-readable file, using something like serde or bincode.
Is the idea okay? Do you have any suggestion on this?

**Comment by saschanaz** [comment]
Looks like the conversion to HashMap takes quite some time, because it hashes things, which is inherently expensive! It might be worth to:

1. Try [BTreeMap](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html) instead, which does not hash?
2. Try using `&str` as key/value instead of String, which requires additional memory copy.

**Reply by Rohupt** [reply]
Thanks! Actually just changing from debug to release decimates the reading time, but I'll try to see if BTreeMap can reduce it even further. Also... I met ownership/borrowing hell when trying to do things with `&str`, so they're gonna stay as `String`s for a while.
